### PR TITLE
Remote file system cache

### PR DESCRIPTION
**Pull request by acquamarin**
# Description

This PR introduces the file cache mechanism for files hosted on http. Users can set the env: HTTP_CACHE_FILE to true to enable the cache. When the user sends a request to the file hosted on the remote file system, kuzu will download the file and cache on local disk.

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ Catalog::Catalog() {
 
 Catalog::Catalog(std::string directory, VirtualFileSystem* fs) {
     if (fs->fileOrPathExists(
-            StorageUtils::getCatalogFilePath(fs, directory, FileVersionType::ORIGINAL))) {
+            StorageUtils::getCatalogFilePath(fs, directory, FileVersionType::ORIGINAL), nullptr)) {
```
Comment (ray6080) [inline src/catalog/catalog.cpp:45]: I think you can omit `nullptr` here as it is the default value.

File: extension/httpfs/src/include/http_config.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "main/client_context.h"
+
+namespace kuzu {
+namespace httpfs {
+
+struct HTTPConfig {
+    explicit HTTPConfig(main::ClientContext* context);
+
+    bool cacheInMemory;
```
Comment (ray6080) [inline extension/httpfs/src/include/http_config.h:11]: I think you mean `cacheToLocalFile` here?

File: extension/httpfs/src/httpfs.cpp [hunk 3]
```diff
@@ -111,8 +124,9 @@ void HTTPFileInfo::initializeClient() {
 }
 
 std::unique_ptr<common::FileInfo> HTTPFileSystem::openFile(const std::string& path, int flags,
-    main::ClientContext* /*context*/, common::FileLockType /*lock_type*/) {
-    auto httpFileInfo = std::make_unique<HTTPFileInfo>(path, this, flags);
+    main::ClientContext* context, common::FileLockType /*lock_type*/) {
+    initCachedFileManager(context);
```
Comment (ray6080) [inline extension/httpfs/src/httpfs.cpp:128]: `initCachedFileManager` should be thread safe.

**Review by ray6080 — APPROVED** [review]

File: extension/httpfs/src/cached_file_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "httpfs.h"
+
+namespace kuzu {
+namespace httpfs {
+
+using namespace common;
+
+CachedFileManager::CachedFileManager(main::ClientContext* context) : vfs{context->getVFSUnsafe()} {
+    cacheDir = context->getDatabasePath() + "/cached_files";
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply extension/httpfs/src/cached_file_manager.cpp:11]: agreed, we can store in an invisible directory.

**Pull request merged**